### PR TITLE
refactor(libsinsp): remove cri extern variables

### PR DESCRIPTION
**Pull request by therealbobo**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

/kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

This removes the extern cri variables in attempt to make libsinsp more maintainable.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by incertum** [comment]
@therealbobo starting some better unit tests for the container engine, will tag you. We should add some more tests as we change things around in the container engine going forward.

File: userspace/libsinsp/cri_settings.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	m_cri_size_timeout(10000),
+	m_cri_runtime_type(CT_CRI),
+	m_cri_unix_socket_path(),
+	m_cri_extra_queries(true)
+{ }
+
+settings::~settings()
+{ }
+
+settings* settings::s_instance = nullptr;
```
Comment (incertum) [inline userspace/libsinsp/cri_settings.cpp:38]: Same feedback as other PR, do we need raw pointers? If possible smart pointers going forward in libs, we have bee burnt a few times.

File: userspace/libsinsp/cri.h [hunk 2]
```diff
@@ -41,13 +41,97 @@ limitations under the License.
 namespace libsinsp {
 namespace cri {
 
-// these shouldn't be globals but we still need references to *the* CRI runtime
-extern std::vector<std::string> s_cri_unix_socket_paths;
-extern int64_t s_cri_timeout;
-// TODO: drop these 2 below
```
Comment (incertum) [inline userspace/libsinsp/cri.h:47]: Drop these 2 below? Anyone knows what this todo item actually means? Can we address it?

File: userspace/libsinsp/cri.h [hunk 3]
```diff
@@ -41,13 +41,97 @@ limitations under the License.
...
-extern std::string s_cri_unix_socket_path;
-extern sinsp_container_type s_cri_runtime_type;
-extern bool s_cri_extra_queries;
+class settings
```
Comment (incertum) [inline userspace/libsinsp/cri.h:44]: Personally would be a fan of naming it  `cri_settings`

File: userspace/libsinsp/container_engine/cri.cpp [hunk 4]
```diff
@@ -57,14 +57,15 @@ constexpr const cgroup_layout CRI_CGROUP_LAYOUT[] = {
 
 cri::cri(container_cache_interface &cache) : container_engine_base(cache)
 {
-	if (s_cri_unix_socket_paths.empty())
+	libsinsp::cri::settings* cri_settings = libsinsp::cri::settings::instance();
+	if (cri_settings->get_cri_unix_socket_paths().empty())
```
Comment (incertum) [inline userspace/libsinsp/container_engine/cri.cpp:61]: Could we add just a tiny new test to make sure the default invocation adds all default socket paths (and no one breaks it in the future)? The new logic would then also go through the LeakSanitizer test in our CI.

File: userspace/libsinsp/cri.h (hunk 2 above)
Comment (therealbobo) [reply userspace/libsinsp/cri.h:47]: Taking a look around,  they could have a legitimate use given that they gives useful information about the actual socket path used by the container runtime and its type. In my opinion, we should keep them

File: userspace/libsinsp/test/cri_settings.ut.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+*/
+
+#include <gtest/gtest.h>
+#include <cri.h>
+#include <cri.hpp>
+#include "./sinsp_with_test_input.h"
+
+
+TEST_F(sinsp_with_test_input, default_cri_socket_paths)
```
Comment (incertum) [inline userspace/libsinsp/test/cri_settings.ut.cpp:25]: Test LGTM! Thanks!

File: userspace/libsinsp/cri_settings.cpp (hunk 1 above)
Comment (therealbobo) [reply userspace/libsinsp/cri_settings.cpp:38]: The problem here is that I didn't take the static local variable approach (like Meyer's Singleton) but, instead, I went for a more straightforward singleton with a static variable in the global scope. Do you know if the static local variable approach guarantees the uniqueness of the object in all the cases? 🤔

Comment (incertum) [reply userspace/libsinsp/cri_settings.cpp:38]: Thanks for the additional comment @therealbobo let's collectively check what the best approach would be here. Maybe the other maintainers have preferences as well.

Comment (incertum) [reply userspace/libsinsp/cri_settings.cpp:38]: Thought more about it and I would prefer a static `get()` function that returns a reference to a `std::unique_ptr` similar to the existing `sinsp_dns_manager`. Reasons being that we adopt a more consistent style throughout the code base and we discussed in previous core maintainers meetings that we won't accept raw pointers anymore going forward unless absolutely necessary, which I believe we don't have a strong case here for an exception. WDYT?

File: userspace/libsinsp/test/cri_settings.ut.cpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include <gtest/gtest.h>
```
Comment (incertum) [inline userspace/libsinsp/test/cri_settings.ut.cpp:19]: Not sure if the CI failures are related to that, but I recall there were hiccups including the cri.hpp, e.g. see https://github.com/falcosecurity/libs/blob/ad3e824cdff5e5e68fe7b95d51cd8f64ee9dc915/userspace/libsinsp/test/container_engine/container_parser_cri_containerd.ut.cpp#L19

**Review by incertum — COMMENTED** [review]
Thanks @therealbobo LGTM, now just need to overcome the CI.

**Review by incertum — APPROVED** [review]
/approve

Thanks @therealbobo!

**Pull request merged**